src/deluge/hid/led/indicator_leds.cpp

```diff
@@ -178,15 +179,15 @@ void setMeterLevel(uint8_t whichKnob, uint8_t level) {
 
 // Level is out of 128
 // Set level and block metering for 500ms
-void setKnobIndicatorLevel(uint8_t whichKnob, uint8_t level) {
+void setKnobIndicatorLevel(uint8_t whichKnob, uint8_t level, bool isBipolar) {
 	if (whichKnob == whichKnobMetering) {
 		uiTimerManager.setTimer(TimerName::METER_INDICATOR_BLINK, 500);
 	}
-	actuallySetKnobIndicatorLevel(whichKnob, level);
+	actuallySetKnobIndicatorLevel(whichKnob, level, isBipolar);
 }
 
 // Just set level
-void actuallySetKnobIndicatorLevel(uint8_t whichKnob, uint8_t level) {
+void actuallySetKnobIndicatorLevel(uint8_t whichKnob, uint8_t level, bool isBipolar) {
```

Comment:
This function can be massively simplified by calculating brightness before the loop, as in the original code

level >> 5 (e.g. divide by 32 and drop remainder) is the number of leds to light up

(level & 31) << 3 is equivalent to (level - (numLEDs * 32)) << 3, and is the remaining brightness (e.g. the remainder of the division)

Then you just need some simple logic to convert negative bipolar values to positive ones before feeding them into the function


As is this change makes the code wildly slower and it runs pretty often, so if we want to deliberately make it less efficient I'd like to see some benchmarks first

Reply:
a bit confused because I am still calculating brightness before the loop. 

As before I am calculating brightnessOutputValue in the loop.

the brightnessOutputValue depends on the indicator being looked at.

level >> 5 is useless for bipolar as it calculates max # of LED's lit up for non-bipolar